File: WinForms/VB/EditorsMainDemo/Modules/ModulePictureEdit.vb
Imports System
Imports System.ComponentModel
Imports System.Drawing
' <comboBoxSizeMode>
Imports DevExpress.XtraEditors.Controls
' </comboBoxSizeMode>
Imports System.Windows.Forms
Imports System.Drawing.Drawing2D
Imports DevExpress.Utils
Imports DevExpress.XtraEditors.ImageEditor
Imports DevExpress.XtraTab
Imports DevExpress.XtraBars.Navigation
Imports DevExpress.Data.Utils

Namespace DevExpress.XtraEditors.Demos

    Public Partial Class ModulePictureEdit
        Inherits TutorialControl

        Public Sub New()
            InitializeComponent()
            If tabPane1 IsNot Nothing Then AddHandler tabPane1.SelectedPageChanged, AddressOf OnSelectedPageChanged
            AddHandler peEditable.ImageEditorDialogShowing, AddressOf PeEditable_ImageEditorDialogShowing
        End Sub

        Protected Overrides ReadOnly Property WhatsThisCodeFileNames As String()
            Get
                Return New String() {"ModulePictureEdit"}
            End Get
        End Property

        Protected Overrides ReadOnly Property WhatsThisXMLFileName As String
            Get
                Return "pictureEdit"
            End Get
        End Property

        Private Sub OnSelectedPageChanged(ByVal sender As Object, ByVal e As SelectedPageChangedEventArgs)
            CalcContentCore()
        End Sub

        <DesignerSerializationVisibility(DesignerSerializationVisibility.Hidden), Browsable(False), EditorBrowsable(EditorBrowsableState.Never)>
        Public Overrides ReadOnly Property ContainsTabs As Boolean
            Get
                Return True
            End Get
        End Property

        Public Overrides Sub SelectTabByName(ByVal tabName As String)
            If String.IsNullOrEmpty(tabName) Then Return
            For Each page As TabNavigationPage In tabPane1.Pages
                If String.Equals(page.Caption.Replace(" ", ""), tabName, StringComparison.OrdinalIgnoreCase) Then
                    tabPane1.SelectedPage = page
                    Exit For
                End If
            Next
        End Sub

        Public Overrides Function GetSelectedTabFullName() As String
            Return tabPane1.SelectedPage.Caption
        End Function

        Protected Overrides Sub UnsubscribeOnEvents()
            MyBase.UnsubscribeOnEvents()
            If tabPane1 IsNot Nothing Then RemoveHandler tabPane1.SelectedPageChanged, AddressOf OnSelectedPageChanged
        End Sub

        Protected Overrides Sub UpdateTabControlPosition(ByVal tabControl As XtraTabControl)
        End Sub

        Protected Overrides Sub CalcContentCore()
            CalcLayoutControlBounds(layoutControl1)
            CalcLayoutControlBounds(layoutControl2)
            CalcLayoutControlBounds(layoutControl4)
        End Sub

        Private updateValues As Boolean = False

        Private aligns As ContentAlignment() = New ContentAlignment() {ContentAlignment.TopLeft, ContentAlignment.TopCenter, ContentAlignment.TopRight, ContentAlignment.MiddleLeft, ContentAlignment.MiddleCenter, ContentAlignment.MiddleRight, ContentAlignment.BottomLeft, ContentAlignment.BottomCenter, ContentAlignment.BottomRight}

        Private Sub ModulePictureEdit_Load(ByVal sender As Object, ByVal e As EventArgs)
            For Each mode As InterpolationMode In [Enum].GetValues(GetType(InterpolationMode))
                If mode <> InterpolationMode.Invalid Then icbInterpolationMode.Properties.Items.Add(New ImageComboBoxItem(mode.ToString(), mode, -1))
            Next

            icbShowZoomSubMenu.Properties.Items.AddEnum(GetType(DefaultBoolean))
            InitValues()
        End Sub

        Private Sub InitValues()
            updateValues = True
            ceShowScrollBars.Checked = peScrollable.Properties.ShowScrollBars
            ceAllowFocused.Checked = peScrollable.Properties.AllowFocused
            cbAllowMenu.Checked = peScrollable.Properties.ShowMenu
            cbAlignment.EditValue = peScrollable.Properties.PictureAlignment.ToString()
            seZoomPercent.Value = New Decimal(peScrollable.Properties.ZoomPercent)
            ceAllowScroll.Checked = peScrollable.Properties.AllowScrollViaMouseDrag
            icbInterpolationMode.EditValue = peScrollable.Properties.PictureInterpolationMode
            icbShowZoomSubMenu.EditValue = peScrollable.Properties.ShowZoomSubMenu
            ceShowCameraItem.Checked = peScrollable.Properties.ShowCameraMenuItem = CameraMenuItemVisibility.Always
            updateValues = False
            UpdateAllowScrollViaMouseDrag()
        End Sub

        '<ceShowScrollBars>
        Private Sub ceShowScrollBars_CheckedChanged(ByVal sender As Object, ByVal e As EventArgs)
            If updateValues Then Return
            peScrollable.Properties.SizeMode = PictureSizeMode.Clip
            peScrollable.Properties.ShowScrollBars = ceShowScrollBars.Checked
            UpdateAllowScrollViaMouseDrag()
        End Sub

        '</ceShowScrollBars>
        Private Sub UpdateAllowScrollViaMouseDrag()
            ceAllowScroll.Enabled = peScrollable.Properties.ShowScrollBars
        End Sub

        '<ceAllowFocused>
        Private Sub ceAllowFocused_CheckedChanged(ByVal sender As Object, ByVal e As EventArgs)
            If updateValues Then Return
            peScrollable.Properties.AllowFocused = ceAllowFocused.Checked
        End Sub

        '</ceAllowFocused>
        '<cbAlignment>
        Private Sub cbAlignment_SelectedIndexChanged(ByVal sender As Object, ByVal e As EventArgs)
            If updateValues Then Return
            peScrollable.Properties.SizeMode = PictureSizeMode.Clip
            Dim selectedIndex As Integer = cbAlignment.SelectedIndex
            peScrollable.Properties.PictureAlignment = aligns(selectedIndex)
        End Sub

        '</cbAlignment>
        Private Sub seZoomPercent_EditValueChanged(ByVal sender As Object, ByVal e As EventArgs)
            ztbZoomPercent.Value = CInt(seZoomPercent.Value)
        End Sub

        Private Sub ztbZoomPercent_EditValueChanged(ByVal sender As Object, ByVal e As EventArgs)
            If updateValues Then Return
            peScrollable.Properties.SizeMode = PictureSizeMode.Clip
            seZoomPercent.Value = ztbZoomPercent.Value
            peScrollable.Properties.ZoomPercent = ztbZoomPercent.Value
        End Sub

        '<ceAllowScroll>
        Private Sub ceAllowScroll_CheckedChanged(ByVal sender As Object, ByVal e As EventArgs)
            If updateValues Then Return
            peScrollable.Properties.SizeMode = PictureSizeMode.Clip
            peScrollable.Properties.AllowScrollViaMouseDrag = ceAllowScroll.Checked
        End Sub

        '</ceAllowScroll>
        '<icbInterpolationMode>
        Private Sub icbInterpolationMode_SelectedIndexChanged(ByVal sender As Object, ByVal e As EventArgs)
            If updateValues Then Return
            peScrollable.Properties.PictureInterpolationMode = CType(icbInterpolationMode.EditValue, InterpolationMode)
        End Sub

        Private Sub peScrollable_ZoomPercentChanged(ByVal sender As Object, ByVal e As EventArgs)
            updateValues = True
            Try
                seZoomPercent.Value = New Decimal(peScrollable.Properties.ZoomPercent)
                ztbZoomPercent.Value = CInt(peScrollable.Properties.ZoomPercent)
            Finally
                updateValues = False
            End Try
        End Sub

        Private Sub icbShowZoomSubMenu_SelectedIndexChanged(ByVal sender As Object, ByVal e As EventArgs)
            If updateValues Then Return
            peScrollable.Properties.ShowZoomSubMenu = CType(icbShowZoomSubMenu.EditValue, DefaultBoolean)
        End Sub

        Private Sub pictureEditSample_Properties_ContextButtonClick(ByVal sender As Object, ByVal e As ContextItemClickEventArgs)
            ContextItemClick(e)
        End Sub

        Private Sub peScrollable_Properties_ContextButtonClick(ByVal sender As Object, ByVal e As ContextItemClickEventArgs)
            ContextItemClick(e)
        End Sub

        Private Sub ContextItemClick(ByVal e As ContextItemClickEventArgs)
            If Equals(e.Item.Name, "itemDownload") Then
                XtraMessageBox.Show("'Download' item clicked")
            ElseIf Equals(e.Item.Name, "itemRemove") Then
                XtraMessageBox.Show("'Remove' item clicked")
            ElseIf Equals(e.Item.Name, "itemInfo") Then
                Dim args As XtraMessageBoxArgs = New XtraMessageBoxArgs()
                args.AllowHtmlText = DefaultBoolean.True
                args.Text = LicenseToolTipController.GetLicenseInfo($"CC BY-SA 3.0|Cristianv14|devexpress.com/carimage24|Resize and crop")
                AddHandler args.HyperlinkClick, Sub(s, a) SafeProcess.Start(a.Link)
                XtraMessageBox.Show(args)
            End If
        End Sub

        Private Sub ceShowCameraItem_CheckedChanged(ByVal sender As Object, ByVal e As EventArgs)
            If updateValues Then Return
            peScrollable.Properties.ShowCameraMenuItem = If(ceShowCameraItem.Checked, CameraMenuItemVisibility.Always, CameraMenuItemVisibility.Never)
        End Sub

        Private Sub OnMaskTypeChanged(ByVal sender As Object, ByVal e As EventArgs)
            Dim text As String = TryCast(comboBoxEdit1.EditValue, String)
            Select Case text
                Case "Circle"
                    peMasked.Properties.OptionsMask.MaskType = PictureEditMaskType.Circle
                Case "RoundedRect"
                    peMasked.Properties.OptionsMask.MaskType = PictureEditMaskType.RoundedRect
                Case "Custom"
                    peMasked.Properties.OptionsMask.MaskType = PictureEditMaskType.Custom
                Case Else
                    peMasked.Properties.OptionsMask.MaskType = PictureEditMaskType.None
            End Select

            OnMaskTypeChanged(peMasked.Properties.OptionsMask.MaskType)
        End Sub

        Protected Sub OnMaskTypeChanged(ByVal type As PictureEditMaskType)
            textEdit3.Enabled =(type <> PictureEditMaskType.None)
            textEdit1.Enabled = textEdit3.Enabled
            textEdit2.Enabled = textEdit1.Enabled
            comboBoxEdit2.Enabled = textEdit2.Enabled
            seRadius.Enabled =(type = PictureEditMaskType.RoundedRect)
            buttonEdit1.Enabled =(type = PictureEditMaskType.Custom)
        End Sub

        Protected Function GetIntArrayFromEditValue(ByVal editValue As Object) As Integer()
            If editValue Is Nothing Then Return Nothing
            Dim vals As String() = editValue.ToString().Split(New String() {", "}, StringSplitOptions.RemoveEmptyEntries)
            Dim res As Integer() = New Integer(vals.Length - 1) {}
            For i As Integer = 0 To vals.Length - 1
                Dim result As Integer = 0
                If Not Integer.TryParse(vals(i), result) Then Return Nothing
                res(i) = result
            Next

            Return res
        End Function

        Private Sub OnMaskOffsetChanged(ByVal sender As Object, ByVal e As EventArgs)
            Dim val As Integer() = GetIntArrayFromEditValue(textEdit2.EditValue)
            If val Is Nothing OrElse val.Length <> 2 Then Return
            Me.peMasked.Properties.OptionsMask.Offset = New Point(val(0), val(1))
        End Sub

        Private Sub OnMaskSizeChanged(ByVal sender As Object, ByVal e As EventArgs)
            Dim val As Integer() = GetIntArrayFromEditValue(textEdit1.EditValue)
            If val Is Nothing OrElse val.Length <> 2 Then Return
            Dim res As Size = If(val(0) <= 0 OrElse val(1) <= 0, Size.Empty, New Size(val(0), val(1)))
            Me.peMasked.Properties.OptionsMask.Size = res
        End Sub

        Private Sub OnMaskMarginChanged(ByVal sender As Object, ByVal e As EventArgs)
            Dim val As Integer() = GetIntArrayFromEditValue(textEdit3.EditValue)
            If val Is Nothing OrElse val.Length <> 4 Then Return
            Me.peMasked.Properties.OptionsMask.Margin = New Padding(val(0), val(1), val(2), val(3))
        End Sub

        Private Sub OnMaskRadiusChanged(ByVal sender As Object, ByVal e As EventArgs)
            Me.peMasked.Properties.OptionsMask.RectCornerRadius = Decimal.ToInt32(seRadius.Value)
        End Sub

        Private Sub OnMaskLayoutChanged(ByVal sender As Object, ByVal e As EventArgs)
            Me.peMasked.Properties.OptionsMask.MaskLayoutMode = GetMaskLayoutType(comboBoxEdit2.SelectedIndex)
        End Sub

        Protected Function GetMaskLayoutType(ByVal index As Integer) As PictureEditMaskLayoutMode
            Select Case index
                Case 1
                    Return PictureEditMaskLayoutMode.ZoomInside
                Case 2
                    Return PictureEditMaskLayoutMode.Stretch
                Case 3
                    Return PictureEditMaskLayoutMode.BottomCenter
                Case 4
                    Return PictureEditMaskLayoutMode.BottomLeft
                Case 5
                    Return PictureEditMaskLayoutMode.BottomRight
                Case 6
                    Return PictureEditMaskLayoutMode.MiddleCenter
                Case 7
                    Return PictureEditMaskLayoutMode.MiddleLeft
                Case 8
                    Return PictureEditMaskLayoutMode.MiddleRight
                Case 9
                    Return PictureEditMaskLayoutMode.TopCenter
                Case 10
                    Return PictureEditMaskLayoutMode.TopLeft
                Case 11
                    Return PictureEditMaskLayoutMode.TopRight
                Case Else
                    Return PictureEditMaskLayoutMode.Default
            End Select
        End Function

        Private Sub buttonEdit1_ButtonClick(ByVal sender As Object, ByVal e As ButtonPressedEventArgs)
            Dim dlg As OpenFileDialog = New OpenFileDialog()
            dlg.Filter = "BMP (*.bmp)|*.bmp|PNG (*.png)|*.png"
            dlg.Title = "Open"
            If dlg.ShowDialog(Me) = DialogResult.OK AndAlso Not Equals(dlg.FileName, Nothing) AndAlso Not Equals(dlg.FileName, String.Empty) Then
                Try
                    Me.peMasked.Properties.OptionsMask.CustomMask = Image.FromFile(dlg.FileName)
                    buttonEdit1.Text = dlg.FileName
                Catch
                End Try
            End If
        End Sub

        Private Sub btnEdit_Click(ByVal sender As Object, ByVal e As EventArgs)
            peEditable.ShowImageEditorDialog()
        End Sub

        Private Sub PeEditable_ImageEditorDialogShowing(ByVal sender As Object, ByVal e As ImageEditorDialogShowingEventArgs)
            AddHandler e.Form.CustomizeCropOptions, Sub(s, ea) CustomizeCropOptions(s, ea)
        End Sub

        Private Sub CustomizeCropOptions(ByVal sender As Object, ByVal e As CustomizeCropOptionsEventArgs)
            If ceCustomRatios.Checked Then
                Dim widescreen = New AspectRatioInfo(1.777F, "16:9")
                Dim standard = New AspectRatioInfo(1.333F, "4:3")
                e.AspectRatios.Clear()
                e.AspectRatios.Add(widescreen)
                e.AspectRatios.Add(standard)
                e.DefaultAspectRatio = widescreen
            End If
        End Sub

        ' <cbAllowMenu>
        Private Sub cbAllowMenu_CheckedChanged(ByVal sender As Object, ByVal e As EventArgs)
            ' <skip>
            If updateValues Then Return
            ' </skip>
            peScrollable.Properties.ShowMenu = cbAllowMenu.Checked
        End Sub
    ' </cbAllowMenu>
    End Class
End Namespace

File: WinForms/CS/EditorsMainDemo/Modules/ModulePictureEdit.cs
using System;
using System.Collections;
using System.ComponentModel;
using System.Drawing;
// <comboBoxSizeMode>
using DevExpress.XtraEditors.Controls;
// </comboBoxSizeMode>
using System.Windows.Forms;
using System.Drawing.Drawing2D;
using DevExpress.Utils;
using DevExpress.XtraEditors.ImageEditor;
using DevExpress.XtraTab;
using DevExpress.XtraBars.Navigation;
using DevExpress.Data.Utils;

namespace DevExpress.XtraEditors.Demos {
    public partial class ModulePictureEdit : TutorialControl {

        public ModulePictureEdit() {
            InitializeComponent();
            if(tabPane1 != null)
                tabPane1.SelectedPageChanged += OnSelectedPageChanged;

            peEditable.ImageEditorDialogShowing += PeEditable_ImageEditorDialogShowing;
        }
        protected override string[] WhatsThisCodeFileNames { get { return new string[] { "ModulePictureEdit" }; } }
        protected override string WhatsThisXMLFileName { get { return "pictureEdit"; } }
        void OnSelectedPageChanged(object sender, SelectedPageChangedEventArgs e) {
            CalcContentCore();
        }
        [DesignerSerializationVisibility(DesignerSerializationVisibility.Hidden), Browsable(false), EditorBrowsable(EditorBrowsableState.Never)]
        public override bool ContainsTabs { get { return true; } }
        public override void SelectTabByName(string tabName) {
            if(string.IsNullOrEmpty(tabName)) return;
            foreach(TabNavigationPage page in tabPane1.Pages) {
                if(string.Equals(page.Caption.Replace(" ",""), tabName, StringComparison.OrdinalIgnoreCase)) {
                    tabPane1.SelectedPage = page;
                    break;
                }
            }
        }
        public override string GetSelectedTabFullName() { return tabPane1.SelectedPage.Caption; }
        protected override void UnsubscribeOnEvents() {
            base.UnsubscribeOnEvents();
            if(tabPane1 != null)
                tabPane1.SelectedPageChanged -= OnSelectedPageChanged;
        }
        protected override void UpdateTabControlPosition(XtraTab.XtraTabControl tabControl) { }
        protected override void CalcContentCore() {
            CalcLayoutControlBounds(layoutControl1);
            CalcLayoutControlBounds(layoutControl2);
            CalcLayoutControlBounds(layoutControl4);
        }

        bool updateValues = false;
        ContentAlignment[] aligns = new ContentAlignment[] { 
				ContentAlignment.TopLeft, 
				ContentAlignment.TopCenter, 
				ContentAlignment.TopRight, 
				ContentAlignment.MiddleLeft, 
				ContentAlignment.MiddleCenter, 
				ContentAlignment.MiddleRight, 
				ContentAlignment.BottomLeft, 
				ContentAlignment.BottomCenter, 
				ContentAlignment.BottomRight };
        private void ModulePictureEdit_Load(object sender, System.EventArgs e) {
            foreach(InterpolationMode mode in Enum.GetValues(typeof(InterpolationMode))) 
                if(mode != InterpolationMode.Invalid)
                    icbInterpolationMode.Properties.Items.Add(new ImageComboBoxItem(mode.ToString(), mode, -1));
            icbShowZoomSubMenu.Properties.Items.AddEnum(typeof(DefaultBoolean));
            InitValues();
        }


        void InitValues() {
            updateValues = true;
            ceShowScrollBars.Checked = peScrollable.Properties.ShowScrollBars;
            ceAllowFocused.Checked = peScrollable.Properties.AllowFocused;
            cbAllowMenu.Checked = peScrollable.Properties.ShowMenu;
            cbAlignment.EditValue = peScrollable.Properties.PictureAlignment.ToString();
            seZoomPercent.Value = new decimal(peScrollable.Properties.ZoomPercent);
            ceAllowScroll.Checked = peScrollable.Properties.AllowScrollViaMouseDrag;
            icbInterpolationMode.EditValue = peScrollable.Properties.PictureInterpolationMode;
            icbShowZoomSubMenu.EditValue = peScrollable.Properties.ShowZoomSubMenu;
            ceShowCameraItem.Checked = peScrollable.Properties.ShowCameraMenuItem == CameraMenuItemVisibility.Always;
            updateValues = false;
            UpdateAllowScrollViaMouseDrag();
        }

        //<ceShowScrollBars>
        private void ceShowScrollBars_CheckedChanged(object sender, EventArgs e) {
            if(updateValues) return;
            peScrollable.Properties.SizeMode = PictureSizeMode.Clip;
            peScrollable.Properties.ShowScrollBars = ceShowScrollBars.Checked;
            UpdateAllowScrollViaMouseDrag();
        }
        //</ceShowScrollBars>
        void UpdateAllowScrollViaMouseDrag() {
            ceAllowScroll.Enabled = peScrollable.Properties.ShowScrollBars;
        }

        //<ceAllowFocused>
        private void ceAllowFocused_CheckedChanged(object sender, EventArgs e) {
            if(updateValues) return;
            peScrollable.Properties.AllowFocused = ceAllowFocused.Checked;
        }
        //</ceAllowFocused>
        //<cbAlignment>
        private void cbAlignment_SelectedIndexChanged(object sender, EventArgs e) {
            if(updateValues) return;
            peScrollable.Properties.SizeMode = PictureSizeMode.Clip;
            int selectedIndex = cbAlignment.SelectedIndex;
            peScrollable.Properties.PictureAlignment = aligns[selectedIndex];
        }
        //</cbAlignment>
        private void seZoomPercent_EditValueChanged(object sender, EventArgs e) {
            ztbZoomPercent.Value = (int)seZoomPercent.Value;
        }

        private void ztbZoomPercent_EditValueChanged(object sender, EventArgs e) {
            if(updateValues) return;
            peScrollable.Properties.SizeMode = PictureSizeMode.Clip;
            seZoomPercent.Value = ztbZoomPercent.Value;
            peScrollable.Properties.ZoomPercent = ztbZoomPercent.Value;
        }
        //<ceAllowScroll>
        private void ceAllowScroll_CheckedChanged(object sender, EventArgs e) {
            if(updateValues) return;
            peScrollable.Properties.SizeMode = PictureSizeMode.Clip;
            peScrollable.Properties.AllowScrollViaMouseDrag = ceAllowScroll.Checked;
        }
        //</ceAllowScroll>
        //<icbInterpolationMode>
        private void icbInterpolationMode_SelectedIndexChanged(object sender, EventArgs e) {
            if(updateValues) return;
            peScrollable.Properties.PictureInterpolationMode = (InterpolationMode)icbInterpolationMode.EditValue;
        }

        private void peScrollable_ZoomPercentChanged(object sender, EventArgs e) {
            updateValues = true;
            try {
                seZoomPercent.Value = new decimal(peScrollable.Properties.ZoomPercent);
                ztbZoomPercent.Value = (int)peScrollable.Properties.ZoomPercent;
            } finally {
                updateValues = false;
            }
        }

        private void icbShowZoomSubMenu_SelectedIndexChanged(object sender, EventArgs e) {
            if(updateValues) return;
            peScrollable.Properties.ShowZoomSubMenu = (DefaultBoolean)icbShowZoomSubMenu.EditValue;
        }

        private void pictureEditSample_Properties_ContextButtonClick(object sender, ContextItemClickEventArgs e) {
            ContextItemClick(e);
        }

        private void peScrollable_Properties_ContextButtonClick(object sender, ContextItemClickEventArgs e) {
            ContextItemClick(e);
        }
        void ContextItemClick(ContextItemClickEventArgs e) {
            if(e.Item.Name == "itemDownload") {
                XtraMessageBox.Show("'Download' item clicked");
            }
            else if(e.Item.Name == "itemRemove") {
                XtraMessageBox.Show("'Remove' item clicked");
            }
            else if(e.Item.Name == "itemInfo") {
                XtraMessageBoxArgs args = new XtraMessageBoxArgs();
                args.AllowHtmlText = DevExpress.Utils.DefaultBoolean.True;
                args.Text = LicenseToolTipController.GetLicenseInfo($"CC BY-SA 3.0|Cristianv14|devexpress.com/carimage24|Resize and crop");
                args.HyperlinkClick += (s, a) => { SafeProcess.Start(a.Link); };
                XtraMessageBox.Show(args);
            }
        }

        private void ceShowCameraItem_CheckedChanged(object sender, EventArgs e) {
            if(updateValues) return;
            peScrollable.Properties.ShowCameraMenuItem = ceShowCameraItem.Checked ? CameraMenuItemVisibility.Always : CameraMenuItemVisibility.Never;
        }
        
        private void OnMaskTypeChanged(object sender, EventArgs e) {
            string text = comboBoxEdit1.EditValue as string;
            switch(text) {
                case "Circle":
                    peMasked.Properties.OptionsMask.MaskType = PictureEditMaskType.Circle;
                    break;
                case "RoundedRect":
                    peMasked.Properties.OptionsMask.MaskType = PictureEditMaskType.RoundedRect;
                    break;
                case "Custom":
                    peMasked.Properties.OptionsMask.MaskType = PictureEditMaskType.Custom;
                    break;
                default:
                    peMasked.Properties.OptionsMask.MaskType = PictureEditMaskType.None;
                    break;
            }
            OnMaskTypeChanged(peMasked.Properties.OptionsMask.MaskType);
        }

        protected void OnMaskTypeChanged(PictureEditMaskType type) {
            comboBoxEdit2.Enabled = textEdit2.Enabled = textEdit1.Enabled = textEdit3.Enabled = (type != PictureEditMaskType.None);
            seRadius.Enabled = (type == PictureEditMaskType.RoundedRect);
            buttonEdit1.Enabled = (type == PictureEditMaskType.Custom);
        }
        protected int[] GetIntArrayFromEditValue(object editValue) {
            if(editValue == null)
                return null;
            string[] vals = editValue.ToString().Split(new string[] { ", " }, StringSplitOptions.RemoveEmptyEntries);
            int[] res = new int[vals.Length];
            for(int i = 0; i < vals.Length; i++) {
                int result = 0;
                if(!int.TryParse(vals[i], out result))
                    return null;
                res[i] = result;
            }
            return res;
        }
        private void OnMaskOffsetChanged(object sender, EventArgs e) {
            int[] val = GetIntArrayFromEditValue(textEdit2.EditValue);
            if(val == null || val.Length != 2) return;
            this.peMasked.Properties.OptionsMask.Offset = new Point(val[0], val[1]);
        }
        private void OnMaskSizeChanged(object sender, EventArgs e) {
            int[] val = GetIntArrayFromEditValue(textEdit1.EditValue);
            if(val == null || val.Length != 2) return;
            Size res = (val[0] <= 0 || val[1] <= 0) ? Size.Empty : new Size(val[0], val[1]);
            this.peMasked.Properties.OptionsMask.Size = res;
        }
        private void OnMaskMarginChanged(object sender, EventArgs e) {
            int[] val = GetIntArrayFromEditValue(textEdit3.EditValue);
            if(val == null || val.Length != 4) return;
            this.peMasked.Properties.OptionsMask.Margin = new System.Windows.Forms.Padding(val[0], val[1], val[2], val[3]);
        }
        private void OnMaskRadiusChanged(object sender, EventArgs e) {
            this.peMasked.Properties.OptionsMask.RectCornerRadius = decimal.ToInt32(seRadius.Value);
        }
        private void OnMaskLayoutChanged(object sender, EventArgs e) {
            this.peMasked.Properties.OptionsMask.MaskLayoutMode = GetMaskLayoutType(this.comboBoxEdit2.SelectedIndex);
        }
        protected PictureEditMaskLayoutMode GetMaskLayoutType(int index) {
            switch(index){
                case 1: return PictureEditMaskLayoutMode.ZoomInside;
                case 2: return PictureEditMaskLayoutMode.Stretch;
                case 3: return PictureEditMaskLayoutMode.BottomCenter;
                case 4: return PictureEditMaskLayoutMode.BottomLeft;
                case 5: return PictureEditMaskLayoutMode.BottomRight;
                case 6: return PictureEditMaskLayoutMode.MiddleCenter;
                case 7: return PictureEditMaskLayoutMode.MiddleLeft;
                case 8: return PictureEditMaskLayoutMode.MiddleRight;
                case 9: return PictureEditMaskLayoutMode.TopCenter;
                case 10: return PictureEditMaskLayoutMode.TopLeft;
                case 11: return PictureEditMaskLayoutMode.TopRight;
                default: return PictureEditMaskLayoutMode.Default;
            }
        }
        private void buttonEdit1_ButtonClick(object sender, ButtonPressedEventArgs e) {
            OpenFileDialog dlg = new OpenFileDialog();
            dlg.Filter = "BMP (*.bmp)|*.bmp|PNG (*.png)|*.png";
            dlg.Title = "Open";
            if(dlg.ShowDialog(this) == DialogResult.OK && dlg.FileName != null && dlg.FileName != string.Empty) {
                try {
                    this.peMasked.Properties.OptionsMask.CustomMask = Image.FromFile(dlg.FileName);
                    this.buttonEdit1.Text = dlg.FileName;
                }
                catch { }
            }
        }

        void btnEdit_Click(object sender, EventArgs e) {
            peEditable.ShowImageEditorDialog();
        }

        void PeEditable_ImageEditorDialogShowing(object sender, ImageEditorDialogShowingEventArgs e) {
            e.Form.CustomizeCropOptions += (s, ea) => { CustomizeCropOptions(s, ea); };
        }

        void CustomizeCropOptions(object sender, CustomizeCropOptionsEventArgs e) {
            if(ceCustomRatios.Checked) {
                var widescreen = new AspectRatioInfo(1.777f, "16:9");
                var standard = new AspectRatioInfo(1.333f, "4:3");
                e.AspectRatios.Clear();
                e.AspectRatios.Add(widescreen);
                e.AspectRatios.Add(standard);
                e.DefaultAspectRatio = widescreen;
            }
        }

        // <cbAllowMenu>
        private void cbAllowMenu_CheckedChanged(object sender, EventArgs e) {
            // <skip>
            if(updateValues) return;
            // </skip>
            peScrollable.Properties.ShowMenu = cbAllowMenu.Checked;
        }
        // </cbAllowMenu>
    }
}

